### PR TITLE
Fixed lag between file-sd and dns-sd

### PR DESCRIPTION
**Pull request by jojohappy**
Signed-off-by: jojohappy <sarahdj0917@gmail.com>

## Changes

fixes: #830 

Update the addresses from file SD by using DNS SD resolver when file SD is updated if necessary at once.

PTAL @GiedriusS @R4scal

### CODE REVIEW

File: cmd/thanos/query.go [hunk 1]
```diff
@@ -355,6 +355,7 @@ func runQuery(
 					}
 					fileSDCache.Update(update)
 					stores.Update(ctxUpdate)
+					dnsProvider.Resolve(ctxUpdate, append(fileSDCache.Addresses(), storeAddrs...))
```
Comment (GiedriusS) [inline cmd/thanos/query.go]: This is good but what happens if there are no file SD targets configured? We'd still have to wait, no?

Comment (bwplotka) [reply cmd/thanos/query.go]: There is another one done every X minutes below (:

Comment (GiedriusS) [reply cmd/thanos/query.go]: Ah sorry, I only gave a quick glance and I haven't noticed that, and I thought we were fixing a different kind of problem.

Comment (jojohappy) [reply cmd/thanos/query.go]: > It's not fixing general lag I guess, this PR fixes just the lag of FileSD vs dns.

@bwplotka Could you explain what is the general lag? 

> It's bit hacky
Agree with you. I think it is a simple way to resolve the issue, and I don't find the general lag that you said.

Comment (bwplotka) [reply cmd/thanos/query.go]: Sorry, by general lag I meant the time between each DNS refresh (that is configured by flag). It's kind of flag as we don't watch for changes (as for DNS it's impossible), we poll every X seconds.

**Pull request merged**